File: TangHongyin/trapping_rain_water.py
class Rect(object):
    x = 0
    y = 0

    def __init__(self, x, y):
        self.x = x
        self.y = y


class Solution(object):

    def trap(self, height):
        queue = []
        queue_index = []
        water = 0

        for i in range(len(height)):

            rect = Rect(1, height[i])

            if len(queue) == 0 or height[i] <= queue[-1].y:
                queue.append(rect)
                queue_index.append(i)
            else:
                last_box = 0
                last_index = 0
                part_box = 0
                while len(queue) > 0 and queue[-1].y <= height[i]:
                    last_box = queue[-1].x * queue[-1].y
                    last_index = queue_index[-1]
                    part_box += last_box
                    queue.pop()
                    queue_index.pop()

                if len(queue) > 0:
                    water += (i - queue_index[-1] - 1) * height[i] - part_box
                    queue.append(Rect(i - queue_index[-1], height[i]))
                    queue_index.append(i)
                else:
                    part_box -= last_box
                    water += (i - last_index - 1) * last_box - part_box
                    queue.append(Rect(1, height[i]))
                    queue_index.append(i)
        return water

if __name__ == '__main__':
    height = [2,6,3,8,2,7,2,5,0]
    solution = Solution()
    print solution.trap(height)

File: CuiYanling/Trapping Rain Water.java
public class Solution {
    public int trap(int[] height) {
        int water = 0,len = height.length;
        if(len<=2) return 0;
        int high[] = new int[len]; //记录i~height-1最大的数的索引
        high[len-1] = len-1;
        for(int i=len-2;i>=0;--i)
            if(height[high[i+1]]>=height[i]) high[i] = high[i+1];
            else high[i]=i;
        
        for(int i=0;i<len;){
            for(;i<len-1&&height[i]<=height[i+1];++i);
            if(i>=len-2) break;
            if(height[high[i+1]]<=height[i]){
                for(int j=i+1;j<high[i+1];++j)
                    water+=height[high[i+1]]-height[j];
                i=high[i+1];
            }else{
                int lefth = height[i];
                for(++i;i<len&&height[i]<=lefth;++i)
                    water+=lefth-height[i];
            }
        }
        return water;
    }
}
